src/stream/src/from_proto/source.rs

```diff
@@ -124,18 +128,61 @@ impl ExecutorBuilder for SourceExecutorBuilder {
                     source_ctrl_opts,
                 )?))
             } else {
-                Ok(Box::new(SourceExecutor::new(
-                    params.actor_context,
-                    schema,
-                    params.pk_indices,
+                let source_exec = SourceExecutor::new(
+                    params.actor_context.clone(),
+                    schema.clone(),
+                    params.pk_indices.clone(),
                     Some(stream_source_core),
-                    params.executor_stats,
+                    params.executor_stats.clone(),
                     barrier_receiver,
                     system_params,
                     params.executor_id,
-                    source_ctrl_opts,
+                    source_ctrl_opts.clone(),
                     params.env.connector_params(),
-                )))
+                );
+
+                let table_type = ExternalTableType::from_properties(&source.properties);
+                if table_type.can_backfill() && let Some(table_desc) = source_info.upstream_table.clone() {
+                    let upstream_table_name = SchemaTableName::from_properties(&source.properties);
+                    let pk_indices = table_desc
+                        .pk
+                        .iter()
+                        .map(|k| k.column_index as usize)
+                        .collect_vec();
+
+                    let order_types = table_desc
+                        .pk
+                        .iter()
+                        .map(|desc| OrderType::from_protobuf(desc.get_order_type().unwrap()))
+                        .collect_vec();
+
+                    let table_reader = table_type.create_table_reader(source.properties.clone(), schema.clone())?;
+                    let external_table = ExternalStorageTable::new(
```

Comment:
Hmm, it looks weird to modify actor graph when building from proto. Consider doing this in stream fragmentor?

https://github.com/risingwavelabs/risingwave/blob/4c0f9f2ed4d8b679887742b831c7ac848af664dc/src/meta/src/stream/stream_graph/actor.rs#L119-L127

Reply:
Do you mean that this is a temporal solution and we'll finally make `CdcBackfill` a standalone executor?

Reply:
Yes